---
title: "Converting 10X sc-RNAseq to Tumor Deconvolution Benchmarking Data"
author: "Tom Snir"
date: "February 11, 2019"
output: html_document
---

The following code will convert 10X sc-RNAseq data to files required to benchmark tumor deconvolution algorithms. To make things clearer, I go through several steps:

- Data cleaning and QC on the sc-RNAseq data
- Dimensionality reduction and clustering, finding differentially expressed genes.
- Extract from the clustering the type of data expected by Tumor Deconvolution algorithms
- Saving the files with the appropriate names and formats

The code below is used on a public 10X dataset, found at:  https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz

#Part 1, loading and cleaning the sc-RNAseq data.

Public 10X data sets can be downloaded as a spare matrix (.mtx) with two additional files, barcodes.tsv and features.tsv (sometimes called genes), each a vector with the names of the cell barcodes and features/genes. The Seurat package which I'm using expects a folder with all 3 files present. Note that there is a raw version of the data and a filtered version, I am using the filtered version to avoid unnecessary complications.

```{r setup}
library(Seurat)
library(dplyr)
library(Matrix)
setwd("C:/Users/Tom Snir/Desktop/Dani/10X/3k_healthy/filtered_gene_bc_matrices/hg19")
```

Read the data and convert it to a Seurat Object, including only cells with at least 200 genes and genes expressed in at least 3 cells (~0.1% of the total number of cells)

```{r}
data.3k <- Read10X("C:/Users/Tom Snir/Desktop/Dani/10X/3k_healthy/filtered_gene_bc_matrices/hg19")

data.1 <- CreateSeuratObject(raw.data = data.3k,
                                          min.cells = 3,
                                          min.genes = 200,
                                          project = "healthy_10K")
```

Looking at the number of genes and UMIs, I decide to remove a few cells with unlikely values.

```{r}
#before QC

VlnPlot(object = data.1,
        features.plot = c("nGene", "nUMI"),
        nCol = 2)

GenePlot(object = data.1, gene1 = "nUMI", gene2 = "nGene", pch.use = '.') 


#QC
data.2 <- FilterCells(object = data.1,
                         subset.names = c("nGene", "nUMI"),
                         high.thresholds = c(2500, 15000))



#after QC

VlnPlot(object = data.2,
        features.plot = c("nGene", "nUMI"),
        nCol = 2)

GenePlot(object = data.2, gene1 = "nUMI", gene2 = "nGene", pch.use = '.') 
```

Normalize the data:

```{r}
#expression before normalization

hist(colSums(data.2@data),
     breaks = 100,
     main = "Total expression before normalisation",
     xlab = "Sum of expression")

#normalize
data.3 <- NormalizeData(object = data.2,
                           normalization.method = "LogNormalize",
                           scale.factor = 1e4)

#expression after normalization

hist(colSums(data.3@data),
     breaks = 100,
     main = "Total expression after normalisation",
     xlab = "Sum of expression")
```

Find variable genes and remove confounders

```{r}
data.3 <- FindVariableGenes(
  object = data.3,
  mean.function = ExpMean, 
  dispersion.function = LogVMR, 
  x.low.cutoff = 0.0125, 
  x.high.cutoff = 3, 
  y.cutoff = 0.5
)

length(data.3@var.genes)

#remove confounders

data.3 <- ScaleData(
  object = data.3, 
  vars.to.regress = c("nUMI")
)

```

#Part 2, Dimensionality reduction and clustering, finding differentially expressed genes.

PCA

```{r}

data.3 <- RunPCA(
  object = data.3, 
  pc.genes = data.3@var.genes, 
  do.print = TRUE, 
  pcs.print = 1:5, 
  genes.print = 5
)


VizPCA(object = data.3, pcs.use = 1:5)

PCAPlot(object = data.3, dim.1 = 1, dim.2 = 2)


PCHeatmap(
  object = data.3, 
  pc.use = 1:5, 
  cells.use = 500, 
  do.balanced = TRUE, 
  label.columns = FALSE,
  use.full = FALSE
)

PCElbowPlot(object = data.3)
```

Find clusters

```{r}
data.3 <- FindClusters(
  object = data.3, 
  reduction.type = "pca", 
  dims.use = 1:10, 
  resolution = 0.6, 
  print.output = 0, 
  save.SNN = TRUE
)

PrintFindClustersParams(object = data.3)

```

tSNE

```{r}
data.3 <- RunTSNE(
  object = data.3,
  dims.use = 1:10,
  do.fast = TRUE
)

TSNEPlot(object = data.3, do.label = TRUE)
```

#Step 3, Extract from the clustering the type of data expected by Tumor Deconvolution algorithms

Tumor deconvolution methods expect several files, which we will create from the data. These files are:
- pure: a matrix of genes as rows and pure tissues/cell types as columns. This is the ground truth for the algorithms.
- mix: several mixes of the pure tissues, with the exact percentages hidden.
- weights: the percentages of the mixes, to be used to evaluate the algorithm
- gene markers: genes used to identify each tissue/cell type.

I will use commands from the Seurat package to create data according to the above.

For the pure tissues (cell populations in our case) I will first look at how many cells make each cluster:
```{r}
table(data.3@ident) #how many cells in each cluster
prop.table(x = table(data.3@ident)) #proportion of each cell
```

For now, I will only continue with the top 5 clusters, meaning I will have 5 pure cell populations.

```{r}
# subsetting cells by cluster
clust0 <- as.matrix(x = data.3@raw.data[, WhichCells(object = data.3, ident = "0")])
clust1 <- as.matrix(x = data.3@raw.data[, WhichCells(object = data.3, ident = "1")])
clust2 <- as.matrix(x = data.3@raw.data[, WhichCells(object = data.3, ident = "2")])
clust3 <- as.matrix(x = data.3@raw.data[, WhichCells(object = data.3, ident = "3")])
clust4 <- as.matrix(x = data.3@raw.data[, WhichCells(object = data.3, ident = "4")])

cluster.averages <- AverageExpression(object = data.3)

top5.averages <- cluster.averages[, (1:5)] #this will be the pure file.

```

Next I will create the weights file, which I will basically determine myself. Later the mix file will be created according to this. the first 5 mixes will have 60% of a pure tissue, and 10% of each remaining tissue. The next 4 mixes will have 35% of two tissues, and 10% of each remaining tissue. Finally, the last mix will have equal amounts of all tissues (20%):

0	  1	  2	  3	  4
60	10	10	10	10
10	60	10	10	10
10	10	60	10	10
10	10	10	60	10
10	10	10	10	60
35	35	10	10	10
10	35	35	10	10
10	10	35	35	10
10	10	10	35	35
20	20	20	20	20


```{r}
weights <- data.frame(
    mix1 = c(60, 10, 10, 10, 10),
    mix2 = c(10, 60, 10, 10, 10),
    mix3 = c(10, 10, 60, 10, 60),
    mix4 = c(10, 10, 10, 60, 10),
    mix5 = c(10, 10, 10, 10, 60),
    mix6 = c(35, 35, 10, 10, 10),
    mix7 = c(10, 35, 35, 10, 10),
    mix8 = c(10, 10, 35, 35, 10),
    mix9 = c(10, 10, 10, 35, 35),
    mix10 = c(20, 20, 20, 20, 20)
    )
```

Next is the actual mix file, which i will create by sampling a proportional number of cells according to the weights file.

```{r}
#first 5 mixes, 60% and 4X10%

mix1_60 <- SubsetData(object = data.3, ident.use = "0", max.cells.per.ident = 600)
mix1_10 <- SubsetData(object = data.3, ident.use = c("1","2","3","4"), max.cells.per.ident = 100)
mix1 <- MergeSeurat(mix1_10, mix1_60)

mix2_60 <- SubsetData(object = data.3, ident.use = "1", max.cells.per.ident = 600)
mix2_10 <- SubsetData(object = data.3, ident.use = c("0","2","3","4"), max.cells.per.ident = 100)
mix2 <- MergeSeurat(mix2_10, mix2_60)

mix3_60 <- SubsetData(object = data.3, ident.use = "2", max.cells.per.ident = 600)
mix3_10 <- SubsetData(object = data.3, ident.use = c("0","1","3","4"), max.cells.per.ident = 100)
mix3 <- MergeSeurat(mix3_10, mix3_60)

mix4_60 <- SubsetData(object = data.3, ident.use = "3", max.cells.per.ident = 600)
mix4_10 <- SubsetData(object = data.3, ident.use = c("0","1","2","4"), max.cells.per.ident = 100)
mix4 <- MergeSeurat(mix4_10, mix4_60)

mix5_60 <- SubsetData(object = data.3, ident.use = "4", max.cells.per.ident = 600)
mix5_10 <- SubsetData(object = data.3, ident.use = c("0","1","2","3"), max.cells.per.ident = 100)
mix5 <- MergeSeurat(mix5_10, mix5_60)

# next 4 mixes, 2X35%, 3X10%

mix6_35 <- SubsetData(object = data.3, ident.use = c("0","1"), max.cells.per.ident = 350)
mix6_10 <- SubsetData(object = data.3, ident.use = c("2","3","4"), max.cells.per.ident = 100)
mix6 <- MergeSeurat(mix1_10, mix1_60)

mix7_35 <- SubsetData(object = data.3, ident.use = c("1","2"), max.cells.per.ident = 350)
mix7_10 <- SubsetData(object = data.3, ident.use = c("0","3","4"), max.cells.per.ident = 100)
mix7 <- MergeSeurat(mix2_10, mix2_60)

mix8_35 <- SubsetData(object = data.3, ident.use = c("2","3"), max.cells.per.ident = 350)
mix8_10 <- SubsetData(object = data.3, ident.use = c("0","1","4"), max.cells.per.ident = 100)
mix8 <- MergeSeurat(mix3_10, mix3_60)

mix9_35 <- SubsetData(object = data.3, ident.use = c("3","4"), max.cells.per.ident = 350)
mix9_10 <- SubsetData(object = data.3, ident.use = c("0","1","2"), max.cells.per.ident = 100)
mix9 <- MergeSeurat(mix4_10, mix4_60)

# final, equal mix

mix10_20 <- SubsetData(object = data.3, ident.use = c("0","1","2","3","4"), max.cells.per.ident = 200)
mix10 <- CreateSeuratObject(mix10_20@data)

#and finally, create averaged-out expression for each mix

mix1.avg <- AverageExpression(object = mix1)
mix2.avg <- AverageExpression(object = mix2)
mix3.avg <- AverageExpression(object = mix3)
mix4.avg <- AverageExpression(object = mix4)
mix5.avg <- AverageExpression(object = mix5)
mix6.avg <- AverageExpression(object = mix6)
mix7.avg <- AverageExpression(object = mix7)
mix8.avg <- AverageExpression(object = mix8)
mix9.avg <- AverageExpression(object = mix9)
mix10.avg <- AverageExpression(object = mix10)

row_names <- rownames(data.3@data)


mix.final <- c(mix1.avg, mix2.avg, mix3.avg, mix4.avg, mix5.avg,
               mix6.avg, mix7.avg, mix8.avg, mix9.avg, mix10.avg)

#need to add gene names to the mix.final

```

And finally, I will find marker genes by looking for differentially expressed genes within each cluster. 

```{r}
all_de_genes <- FindAllMarkers(data.3, test.use = "bimod", do.print = TRUE)
```

Look at the markers we found:

```{r}
# Select markers for plotting on a Heatmap (positive markers with high discriminatory power)
markers.use <- subset(all_de_genes, avg_logFC>0 & pct.1 > 0.8)$gene

# Draw a heatmap of all cells for these marker genes
DoHeatmap(data.3, genes.use = markers.use, slim.col.label = TRUE, remove.key = TRUE, cex.row = 0.1)
```

#Part 4, Saving the files with the appropriate names and formats

```{r}
#Pure:
write.table(top5.averages, "pure.tsv", quote=FALSE, sep='\t', col.names = FALSE) 

#Weights:
write.table(weights, "weights.tsv", quote=FALSE, sep='\t', col.names = FALSE) 

#Mix:
write.table(mix.final, "mix.tsv", quote=FALSE, sep='\t', col.names = FALSE) 

#Gene_list:
write.table(all_de_genes, "gene_list.tsv", quote=FALSE, sep='\t', col.names = FALSE) 
```

